capreolus/sampler/__init__.py

```diff
@@ -0,0 +1,100 @@
+import random
+import torch.utils.data
+
+from capreolus.utils.loginit import get_logger
+
+
+logger = get_logger(__name__)
+
+
+class TrainDataset(torch.utils.data.IterableDataset):
+    """
+    Samples training data. Intended to be used with a pytorch DataLoader
+    """
+
+    def __init__(self, search_run, benchmark, extractor):
+        self.search_run = search_run
+        self.benchmark = benchmark
+        self.extractor = extractor
+        self.iterations = 0
+
+    def __iter__(self):
+        """
+        Returns: Triplets of the form (query_feature, posdoc_feature, negdoc_feature)
+        """
+        extractor = self.extractor
+        benchmark = self.benchmark
+        search_run = self.search_run
+
+        train_qids = benchmark.folds[benchmark.cfg["fold"]]["train_qids"]
+        is_rundocsonly = benchmark.cfg["rundocsonly"]
+        if is_rundocsonly:
+            # Create qrels from only those docs that showed up in the search results
+            qrels = {
+                qid: {docid: label for docid, label in benchmark.qrels[qid].items() if docid in search_run[qid]}
+                for qid in train_qids
+            }
+        else:
+            qrels = benchmark.qrels
+
+        qid_to_reldocs = {qid: [docid for docid, label in qrels[qid].items() if label > 0] for qid in qrels}
+        qid_to_negdocs = {qid: [docid for docid, label in qrels[qid].items() if label <= 0] for qid in qrels}
+
+        # Remove the qids that does not have relevant/irrelevant docs associated with it
+        valid_qids = [qid for qid in train_qids if qid_to_reldocs.get(qid) and qid_to_negdocs.get(qid)]
+
+        # Convert each query and doc id to the corresponding feature/embedding and yield
+        def genf():
+            while True:
+                random.seed(self.iterations)
```

Comment:
`self.iterations` is the number of times a `yield` was made. This number would be `num_iter * batch_size` in fact. This would be a problem if the batch_size changes between 2 experiments/runs. Is that going to be an issue?

Reply:
That isn't an issue, but I think it doesn't make sense to count inside of the `Dataset`, because this has no advantage over just setting the seed once as we previously. Instead the trainer should set `self.iteration` to reflect the correct iteration, so that the there is no difference between e.g. running the sampler from iteration 1-10 or running it from 1-5 and then later running 5-10. I can modify this to integrate with the trainer though